File: Atcoder_dp/O_Matching.py
I = lambda: input()
II = lambda: int(input())
MII = lambda: map(int, input().split())
LI = lambda: list(input().split())
LII = lambda: list(map(int, input().split()))
LGMII = lambda: map(lambda x: int(x) - 1, input().split())
LGLII = lambda: list(map(lambda x: int(x) - 1, input().split()))

from functools import lru_cache

inf = float('inf')

n = II()
MOD = 10**9+7
a = [LII() for _ in range(n)]

@lru_cache(None)
def go(i,mask):
    if i==n:
        return 1 if mask==(1<<n)-1 else 0
    res = 0
    for j in range(n):
        if (not mask&1<<j ) and a[i][j]:
            res += go(i+1,mask|1<<j)
            res%=MOD
    return res%MOD
print(go(0,0))

File: Atcoder_dp/O_Matching.cpp
#include <iostream>
#include <vector>
using namespace std;

const int MOD = 1e9 + 7;
int n;
vector<vector<int>> a;
vector<int> dp;

int go(int i, int mask) {
    if (i == n) return mask == (1 << n) - 1 ? 1 : 0;
    if (dp[mask] != -1) return dp[mask];
    int res = 0;
    for (int j = 0; j < n; ++j) {
        if (!(mask & (1 << j)) && a[i][j]) {
            res = (res + go(i + 1, mask | (1 << j))) % MOD;
        }
    }
    return dp[mask] = res;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cin >> n;
    a.assign(n, vector<int>(n));
    for (int i = 0; i < n; ++i)
        for (int j = 0; j < n; ++j)
            cin >> a[i][j];
    dp.assign(1 << n, -1);
    cout << go(0, 0) << '\n';
    return 0;
}
